src/AsmResolver.DotNet/Cloning/MemberCloner.Fields.cs

```diff
@@ -39,7 +39,13 @@ private static FieldDefinition CreateFieldStub(MemberCloneContext context, Field
         private void DeepCopyFields(MemberCloneContext context)
         {
             foreach (var field in _fieldsToClone)
+            {
                 DeepCopyField(context, field);
+                var clonedMember = (FieldDefinition)context.ClonedMembers[field];
+                _clonerListener.OnClonedMember(field, clonedMember);
+                if (_clonerListener is MemberClonerListener listener)
```

Comment:
I think we should move this pattern matching (and all similar subsequent matchings) out of the `MemberCloner` class and into `MemberClonerListener::OnClonedMember` (i.e. with a switch case). Hardwiring to a specific implementation of an interface is kind of an anti-pattern. It restricts how the cloner and the listener can be used.

How about we move all `OnClonedXXX` from `MemberClonerListener` into the `IMemberClonerInterface` interface directly, and remove the generic `OnClonedMember` from the interface, but keep it in the abstract class? We can make then everything `virtual` in `MemberClonerListener` and let the default implementations just call `OnClonedMember`. We can then also let `CallbackClonerListener` inherit from the base class instead of the interface. Should get rid of all these `if`s in `MemberCloner`, and also makes it easier to make your own listeners. What do you think?